Add test plan support within Tuist test

stephanecopin opened the pull request:
Resolves https://github.com/tuist/tuist/issues/5080

### Short description 📝

Add 5 new flags to `tuist test`:
- One to use test plan (`--test-plan`)
- Two to specify which configurations to use (`--only-test-configuration`, `--skip-test-configuration`)
- Two to specify which tests to use (`--only-testing`, `--skip-testing`)

The changes are integrated with the Tuist graph & cache system, and will be optimized to build only what's changed.

Open to any suggestions regarding to naming the flags, or if they should be handled differently that I've implemented in the PR.  
Implementation-wide:
1. Would like people knowing how the test caching works to peek at it and let me know if there's anything that looks off. I based my changes mostly by analyzing what the code did and adapting it to work with Test Plans, while maintaining backward compatibility.
2. I've had to create new helpers in multiple places. Let me know if they don't belong in one place or the other and I'll move them.

### How to test the changes locally 🧐

Changes are unit tested and can also be tested using `./fourier test tuist acceptance projects/tuist/features/test-plan.feature`

### Contributor checklist ✅

- [x] The code has been linted using run `./fourier lint tuist --fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

shahzadmajeed commented on Sources/TuistCore/Graph/GraphTraverser.swift:
```diff
@@ -52,6 +52,41 @@ public class GraphTraverser: GraphTraversing {
+        basedOn targets: GraphTargets,
+        testPlan: String?,
+        includedTargets: Set<String>,
+        excludedTargets: Set<String>,
```
Is there a need to check for an intersection between `includedTargets` & `excludedTargets` so that we can throw an early error about misconfiguration?

stephanecopin replied:
Let me know if you should we should still do that considering my comments below.

There could still be errors caught there for configurations that don't make sense, for example if `includedTargets` & `excludedTargets`'s intersection is non-`nil`, as it in this case the user would to include and exclude the same targets. Down to implement this check if we want it

danyf90 replied:
yep, I would implement it in the command

stephanecopin replied:
@danyf90 I've added validation of that to the command

shahzadmajeed commented on projects/docs/docs/commands/test.md:
````diff
@@ -46,15 +46,28 @@ tuist test --device "iPhone X" --os 14.0
+**Test in parallel**
+
+```bash
+tuist test --additional-parameters -parallel-testing-enabled
````
We are not supporting these params yet right?

stephanecopin replied:
Removed

shahzadmajeed commented on projects/docs/docs/commands/test.md:
```diff
@@ -46,15 +46,28 @@ tuist test --device "iPhone X" --os 14.0
+| `--configuration`            | `-C`  | `The configuration to be used when building the scheme.`                                                                               |                   | No       |
+| `--skip-ui-tests`            | n/a   | `When passed, it skips testing UI Tests targets.`                                                                                      | False             | No       |
+| `--result-bundle-path`       | `-T`  | `Path where test result bundle will be saved`                                                                                          |                   | No       |
+| `--retry-count`              | n/a   | `Tests will retry <number> of times until they succeed.`                                                                               | 0                 |          |
```
Is there no short form of new arguments? I thought argument parser creates them automatically?

stephanecopin commented on Sources/TuistCache/Mappers/Graph/FocusTargetsGraphMappers.swift:
```diff
@@ -5,19 +5,33 @@ import TuistGraph
 
 /// `FocusTargetsGraphMappers` is used to filter out some targets and their dependencies and tests targets.
 public final class FocusTargetsGraphMappers: GraphMapping {
+    // When specified, if includedTargets is empty it will automatically all targets in the test plan
```
Done!

stephanecopin commented on Sources/TuistCache/Mappers/Graph/FocusTargetsGraphMappers.swift:
```diff
@@ -32,4 +46,17 @@ public final class FocusTargetsGraphMappers: GraphMapping {
         return (graph, [])
     }
+
+    private func isIncluded(_ target: GraphTarget) -> Bool {
```
Removed 😄

stephanecopin commented on Sources/TuistCache/Mappers/Graph/TestsCacheGraphMapper.swift:
```diff
@@ -145,19 +166,25 @@ public final class TestsCacheGraphMapper: GraphMapping {
+                    visited: &visited,
+                    testsCacheDirectory: testsCacheDirectory
+                )
+            },
```
Done

stephanecopin commented on Sources/TuistAutomation/Utilities/BuildGraphInspector.swift:
```diff
@@ -24,8 +24,15 @@ public protocol BuildGraphInspecting {
+    func testableTarget(
+        scheme: Scheme,
+        testPlan: String?,
+        testTargets: [TestIdentifier],
+        skipTestTargets: [TestIdentifier],
```
So they are actually not mutually exclusive, they can be used together but `only` will take precedence over `skip`. If this didn't read like so, the help message should probably be rewritten so it's clearer - would you have any suggestions?

For example:
Let's consider we have one `TestTarget`, with 3 test cases (`TestCase1`, `TestCase2`, `TestCase3`): if we specify `--test-target TestTarget --skip-test-targets TestTarget/TestCase1/TestFunction TestTarget/TestCase2`, we will test everything in `TestTarget`, except for the test function `TestFunction` in `TestCase1` and `TestCase2`. Everything else will be tested.

danyf90 replied:
I got it now, so it's not a matter of precedence, is just that you apply `only` first, and then `skip`.
If we talk about precedence I understand that `only` overrides `skip`

danyf90 replied:
Can you point me where this logic is implemented?
In `isIncluded` method looks we just check the target name

stephanecopin commented on Sources/TuistKit/Services/TestService.swift:
```diff
@@ -77,7 +95,12 @@ final class TestService {
         resultBundlePath: AbsolutePath?,
-        retryCount: Int
+        retryCount: Int,
+        testPlan: String?,
+        testTargets: [TestIdentifier],
+        skipTestTargets: [TestIdentifier],
+        testConfigurations: [String],
+        skipTestConfigurations: [String]
```
`testTargets` & `skipTestTargets` can be used without a test plan, however `testConfigurations` & `skipTestConfigurations` can only be used when a test plan is specified. I can perhaps refactor them as a `struct` in this case? Like `TestPlanConfiguration` with the 3 fields

stephanecopin commented on Sources/TuistCore/Graph/GraphTraverser.swift:
```diff
@@ -52,6 +52,41 @@ public class GraphTraverser: GraphTraversing {
     }
 
+    public func filterIncludedTargets<GraphTargets: Collection>(
+        basedOn targets: GraphTargets,
```
We could - right now the function may be called with an `Array` or `Set` in the current implementation, so making it generic instead of converting to `Set` made sense to me. Down to change if that's not usually done in the codebase.

qq as well: I think work is still ongoing with Xcode 14.2 & Swift 5.7, but once available would it work to just use `some Collection<GraphTarget>`?

danyf90 replied:
yep, `some Collection<GraphTarget>` would be ideal, let's keep it as it is for now 🙏

stephanecopin commented on Sources/TuistCore/Graph/GraphTraverser.swift:
```diff
@@ -52,6 +52,41 @@ public class GraphTraverser: GraphTraversing {
         allTargets(excludingExternalTargets: true)
     }
 
+    public func filterIncludedTargets<GraphTargets: Collection>(
```
Works for me, will move it in the module.

stephanecopin commented on Sources/TuistKit/Commands/TestCommand.swift:
```diff
@@ -65,6 +66,42 @@ struct TestCommand: AsyncParsableCommand {
+    @Option(
+        name: .long,
+        parsing: .upToNextOption,
+        help: "The list of test identifiers you want to test. Takes precedence over --skip-testing",
```
I mirrored the help from xcodebuild here, and I've added the validation mentioned [here](https://github.com/tuist/tuist/pull/5135#discussion_r1181259538), is there anything more I should do here?
```suggestion
        help: "The list of test identifiers you want to test. Takes precedence over --skip-testing",
```

danyf90 commented on Sources/TuistCache/Mappers/Graph/FocusTargetsGraphMappers.swift:
```diff
@@ -5,19 +5,33 @@ import TuistGraph
+    let excludedTargets: Set<String>
 
-    public init(includedTargets: Set<String>) {
+    public init(
+        testPlan: String? = nil,
+        includedTargets: Set<String>,
+        excludedTargets: Set<String> = []
+    ) {
```
avoid defaults in internal methods, they just risk you forget passing parameters
```suggestion
    public init(testPlan: String?, includedTargets: Set<String>, excludedTargets: Set<String>) {
```

danyf90 commented on Sources/TuistKit/Commands/TestCommand.swift:
```diff
@@ -65,6 +66,49 @@ struct TestCommand: AsyncParsableCommand {
+    @Option(
+        name: .long,
+        parsing: .upToNextOption,
+        help: "The list of test identifiers you want to test. Takes precedence over --skip-testing",
```
```suggestion
        help: "The list of test identifiers you want to test. Expected format is TestTarget[/TestClass[/TestMethod]]. It is applied before --skip-testing",
```

danyf90 commented on Sources/TuistKit/Commands/TestCommand.swift:
```diff
@@ -65,6 +66,49 @@ struct TestCommand: AsyncParsableCommand {
+    @Option(
+        name: .long,
+        parsing: .upToNextOption,
+        help: "The list of test identifiers you want to skip testing.",
```
```suggestion
        help: "The list of test identifiers you want to skip testing. Expected format is TestTarget[/TestClass[/TestMethod]].",
```

danyf90 commented on Sources/TuistKit/Commands/TestCommand.swift:
```diff
@@ -65,6 +66,49 @@ struct TestCommand: AsyncParsableCommand {
+    @Option(
+        name: .long,
+        parsing: .upToNextOption,
+        help: "The list of configurations you want to test. Takes precedence over --skip-test-configuration"
```
```suggestion
        help: "The list of configurations you want to test. It is applied before --skip-test-configuration"
```

danyf90 commented on Sources/TuistKit/Commands/TestCommand.swift:
```diff
@@ -65,6 +66,49 @@ struct TestCommand: AsyncParsableCommand {
+    )
+    var skipTestTargets: [TestIdentifier] = []
+
+    @Option(
+        name: .long,
+        parsing: .upToNextOption,
+        help: "The list of configurations you want to test. Takes precedence over --skip-test-configuration"
+    )
+    var testConfigurations: [String] = []
+
+    @Option(
+        name: .long,
+        parsing: .upToNextOption,
+        help: "The list of configurations you want to skip testing."
+    )
+    var skipTestConfigurations: [String] = []
```
How does it work if you specify both of these?

stephanecopin replied:
Not sure, I'll double check

danyf90 commented on Sources/TuistKit/Commands/TestCommand.swift:
```diff
@@ -65,6 +66,49 @@ struct TestCommand: AsyncParsableCommand {
+    )
+    var skipTestConfigurations: [String] = []
+
+    func validate() throws {
+        let intersection = Set(testTargets).intersection(skipTestTargets)
+        if !intersection.isEmpty {
+            throw ValidationError.invalidTestTargetsOptions(intersection)
+        }
+    }
```
There are also other invalid scenario. For example, if testTargets is [`TargetName`] it is invalid/wrong to have skipTestTarget as [`AnotherTargetName/ClassName`], as it doesn't skip anything from the selected set

stephanecopin replied:
Indeed, though I'm not sure this would be considered an error as technically `AnotherTargetName/ClassName` will be skipped. Maybe we could output some some of warning instead?

danyf90 replied:
It's cleary an error in passing the parameter, better to let the user know by making the command fail

danyf90 commented on projects/docs/docs/commands/test.md:
```diff
@@ -48,13 +48,20 @@ One of the benefits of using Tuist over other automation tools is that developer
+| `--result-bundle-path`       | `-T`  | `Path where test result bundle will be saved`                                                                                          |                   | No       |
+| `--retry-count`              | n/a   | `Tests will retry <number> of times until they succeed.`                                                                               | 0                 |          |
+| `--test-plan`                | n/a   | `When passed, the test plan to run. The scheme parameter is required when a test plan is passed.`                                      |                   | No       |
+| `--test-targets`             | n/a   | `Specify which tests should run in the test-identifier* format. Takes precedence over --skip-testing`                                  |                   | No       |
```
```suggestion
| `--test-targets`             | n/a   | `Specify which tests should run in the test-identifier* format. It is applied before --skip-testing`                                  |                   | No       |
```

danyf90 commented on projects/docs/docs/commands/test.md:
```diff
@@ -48,13 +48,20 @@ One of the benefits of using Tuist over other automation tools is that developer
+| `--test-plan`                | n/a   | `When passed, the test plan to run. The scheme parameter is required when a test plan is passed.`                                      |                   | No       |
+| `--test-targets`             | n/a   | `Specify which tests should run in the test-identifier* format. Takes precedence over --skip-testing`                                  |                   | No       |
+| `--skip-test-targets`        | n/a   | `Specify which tests should not run be passed in the test-identifier* format.`                                                         |                   | No       |
+| `--test-configurations`      | n/a   | `When a test plan is specified, specify which test configurations to use for the run. Takes precedence over --skip-test-configuration` |                   | No       |
```
```suggestion
| `--test-configurations`      | n/a   | `When a test plan is specified, specify which test configurations to use for the run. It is applied before --skip-test-configuration` |                   | No       |
```

pepicrft commented on Tests/TuistCacheTests/Extensions/GraphTraverser+ExtrasTests.swift:
```diff
@@ -0,0 +1,157 @@
+@testable import TuistCache
+
+class GraphTraverserExtrasTests: XCTestCase {
+    func test_filterIncludedTargets_includeExcluded() throws {
```
The name is not very descriptive of the scenario. I'd follow the convention:

```swift
func test_{function_name}_when_{scenario}
```

stephanecopin replied:
Yeah, my intent here was to test all the possible scenarii involved using `filterIncludedTargets`. I can split them up

pepicrft commented on Tests/TuistCacheTests/Extensions/GraphTraverser+ExtrasTests.swift:
```diff
@@ -0,0 +1,157 @@
+        XCTAssertEqual(Set(filteredTargets.map(\.target)), [tests1])
+    }
+
+    func test_filterIncudedTargets_testPlan() throws {
```
Same comment here regarding the name of the scenario.

pepicrft commented on Sources/TuistAutomation/Utilities/BuildGraphInspector.swift:
```diff
@@ -101,9 +108,33 @@ public final class BuildGraphInspector: BuildGraphInspecting {
-    public func testableTarget(scheme: Scheme, graphTraverser: GraphTraversing) -> GraphTarget? {
-        guard let testTarget = scheme.testAction?.targets.first else { return nil }
-        return graphTraverser.target(path: testTarget.target.projectPath, name: testTarget.target.name)
+    public func testableTarget(
+        scheme: Scheme,
+        testPlan: String?,
+        testTargets: [TestIdentifier],
+        skipTestTargets: [TestIdentifier],
+        graphTraverser: GraphTraversing
+    ) -> GraphTarget? {
+        func isIncluded(_ testTarget: TestPlan.TestTarget) -> Bool {
+            if !testTarget.isEnabled {
+                return false
+            } else if testTargets.isEmpty {
+                return !skipTestTargets.contains { $0.target == testTarget.target.name }
+            } else {
+              return testTargets.contains { $0.target == testTarget.target.name }
+            }
+        }
+
+        if let testPlanName = testPlan,
+           let testPlan = scheme.testAction?.testPlans?.first(where: { $0.name == testPlanName }),
+           let target = testPlan.testTargets.first(where: { isIncluded($0) })?.target
+        {
+            return graphTraverser.target(path: target.projectPath, name: target.name)
+        } else if let testTarget = scheme.testAction?.targets.first {
+            return graphTraverser.target(path: testTarget.target.projectPath, name: testTarget.target.name)
+        } else {
+            return nil
+        }
```
I'd add unit tests that cover the various scenarios in this method.

pepicrft commented on Sources/TuistAutomation/Utilities/BuildGraphInspector.swift:
```diff
@@ -122,7 +153,7 @@ public final class BuildGraphInspector: BuildGraphInspecting {
     public func testableSchemes(graphTraverser: GraphTraversing) -> [Scheme] {
         graphTraverser.schemes()
-            .filter { $0.testAction?.targets.isEmpty == false }
+            .filter { $0.testAction?.targets.isEmpty == false || $0.testAction?.testPlans?.isEmpty == false }
```
I'd also add a unit test that covers this one.

pepicrft commented on Sources/TuistCore/Graph/GraphTraverser.swift:
```diff
@@ -52,6 +52,10 @@ public class GraphTraverser: GraphTraversing {
         allTargets(excludingExternalTargets: true)
     }
 
+    public func allTestPlans() -> Set<TestPlan> {
```
I'd add tests for the new functions in this class.

pepicrft commented on Sources/TuistLoader/Models/XCTestPlan.swift:
```diff
@@ -0,0 +1,49 @@
+struct XCTestPlan: Decodable {
```
I'd import `Foundation` here. Is there a reason why this file is not in the `TuistGraph` target where all the other modules live?

pepicrft replied:
I'd also add coddle tests like we are doing with the other models.

stephanecopin replied:
This is an internal object representing the JSON schema for `.xctestplan` files. It's used exclusively within `TuistLoader` (which loads it from the disk), and then `TuistLoader` uses `TuistGraph`'s `TestPlan` to build a Tuist-representation of the test plan (see the file right above).  
Let me know if that makes sense to you, I can add some comments to clarify if needed!

There is discussion to improve that so Test Plan are generated by Tuist directly (so it wouldn't need to be loaded anymore) [here](https://github.com/tuist/tuist/discussions/5214), and I think this should be done eventually but also think it's out of scope for this PR.

pepicrft commented on Sources/TuistLoader/Models+ManifestMappers/TestAction+ManifestMapper.swift:
```diff
@@ -89,3 +89,26 @@ extension TuistGraph.TestAction {
     }
 }
+
+extension TestPlan {
```
Let's be consistent with the other `{Model}+ManifestMapper.swift` and extract this one into a `TestPlan+ManifestMapper.swift` file.

pepicrft reviewed (changes requested):
Thanks for contributing this one @stephanecopin 🙏🏼. I went through it and left a handful of suggestions. Let me know when they are address and I'll give the PR another pass.

stephanecopin commented on Sources/TuistCache/Extensions/GraphTraverser+Extras.swift:
```diff
@@ -0,0 +1,36 @@
+        }
+
+        lazy var allInternalTargets = allInternalTargets().map(\.target.name)
+        return Set(
+            targets.filter { target in
+                if !includedTargets.isEmpty {
+                    return includedTargets.contains(target.target.name)
+                }
+                if excludedTargets.contains(target.target.name) {
+                    return false
+                }
+                if let allTestPlansTargetNames = allTestPlansTargetNames {
+                    return allTestPlansTargetNames.contains(target.target.name)
+                }
+                return excludingExternalTargets ? allInternalTargets.contains(target.target.name) : true
+            }
+        )
+    }
```
Ah, nice catch! Indeed, the logic should be:

_1. If a test plan is specified, if the target target is not in the test plan, exclude it_
2. Include the target target if it's included in in includedTargets.
3. Exclude it if it's included in excludedTargets.
_4. If we're excluding external targets, only include the target if internal. If we're including external targets, include the target_.

Does that make more sense?

As for the name of the function, is it fine with that new logic, and otherwise what would you suggest? That's the best I had 😅

stephanecopin commented on Sources/TuistKit/Services/TestService.swift:
```diff
@@ -10,22 +10,40 @@ import TuistSupport
+            if existing.isEmpty {
+                existingMessage = "No test plans are defined for this scheme"
+            } else {
+                existingMessage = "The available test plans are: \(existing.joined(separator: ","))"
```
Works for me, I'll add the info about the scheme.

As for matching test plan against existing ones using some sort a fuzzy search, I'd rather keep that as a separate improvement as the same logic could be applied in other places (e.g. the scheme above, but probably other places within Tuist), let me know what you think! We can add an improvement issue so we don't forgot about it.

stephanecopin commented on Tests/TuistCacheTests/Extensions/GraphTraverser+ExtrasTests.swift:
```diff
@@ -0,0 +1,157 @@
+        let subject = GraphTraverser(graph: graph)
+
+        var filteredTargets: Set<GraphTarget>
+        // When
+        filteredTargets = subject.filterIncludedTargets(
+            basedOn: subject.allTargets(),
+            testPlan: nil,
+            includedTargets: [tests1.name],
+            excludedTargets: []
+        )
+
+        // Then
+        XCTAssertEqual(filteredTargets.map(\.target), [tests1])
+
+        // When
+        filteredTargets = subject.filterIncludedTargets(
+            basedOn: subject.allTargets(),
+            testPlan: nil,
+            includedTargets: [tests1.name],
+            excludedTargets: [tests1.name]
+        )
```
Works for me! My intent initially was to avoid duplicating the inputs of tests too much. I'll see if I can replace them with properties and not make it too confusing.

stephanecopin reviewed (commented):
Thanks @pepicrft for your review! I've answer to the relevant comments, and will address your feedback. Not sure when I'll be able to, I'll keep you posted!

fortmarek commented on Sources/TuistKit/Commands/TestCommand.swift:
```diff
@@ -65,6 +66,68 @@ struct TestCommand: AsyncParsableCommand {
+    )
+    var skipTestConfigurations: [String] = []
+
+    func validate() throws {
+        let targetsIntersection = Set(testTargets)
+            .intersection(skipTestTargets)
+        if !targetsIntersection.isEmpty {
+            throw ValidationError.duplicatedTestTargets(targetsIntersection)
+        }
+        if !testTargets.isEmpty {
+            // --test-targets Test --skip-test-targets AnotherTest
+            let skipTestTargetsOnly = Set(skipTestTargets.map { TestIdentifier(target: $0.target) })
+            let testTargetsOnly = testTargets.map { TestIdentifier(target: $0.target) }
+            let targetsOnlyIntersection = skipTestTargetsOnly.intersection(testTargetsOnly)
+            if targetsOnlyIntersection.isEmpty {
+                throw ValidationError.nothingToSkip(skipped: skipTestTargets.filter { skipTarget in !testTargetsOnly.contains(TestIdentifier(target: skipTarget.target)) }, included: testTargets)
+            }
+
+            // --test-targets Test/MyTest --skip-test-targets Test/AnotherTest
+            let skipTestTargetsClasses = Set(skipTestTargets.map { TestIdentifier(target: $0.target, class: $0.class) })
+            let testTargetsClasses = testTargets.map { TestIdentifier(target: $0.target, class: $0.class) }
+            let targetsClassesIntersection = skipTestTargetsClasses.intersection(testTargetsClasses)
+                .intersection(testTargetsClasses.map { TestIdentifier(target: $0.target, class: $0.class) })
+            if targetsClassesIntersection.isEmpty {
+                throw ValidationError.nothingToSkip(skipped: skipTestTargets.filter { skipTarget in !testTargetsClasses.contains { $0 == TestIdentifier(target: skipTarget.target, class: skipTarget.class) } }, included: testTargets)
+            }
+        }
+    }
```
`TestService` would most likely be the right place. Or a new utility class.

fortmarek commented on Sources/TuistCache/Extensions/GraphTraverser+Extras.swift:
```diff
@@ -0,0 +1,36 @@
+            allTestPlansTargetNames = nil
+        }
+
+        lazy var allInternalTargets = allInternalTargets().map(\.target.name)
```
Why is this var `lazy`?

stephanecopin replied:
It is lazy as it is not used in all code path, so it's initialized only when used.

fortmarek commented on Sources/TuistCore/Models/TestIdentifier.swift:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public let target: String
+    public let `class`: String?
```
Maybe we should rename this to `className` , so we don't have to escape it everywhere?

fortmarek commented on Sources/TuistCore/Models/TestIdentifier.swift:
```diff
@@ -0,0 +1,58 @@
+        self.method = method
+    }
+
+    public init(string: String) throws {
```
Might be worth testing this init

fortmarek commented on projects/tuist/features/step_definitions/test.rb:
```diff
@@ -1,47 +1,54 @@
 # frozen_string_literal: true
 
-require "xcodeproj"
+require 'xcodeproj'
```
nit: I don't see a reason for the change of quotation marks? Makes it harder to follow the actual changes here.  In the other `step_definitions` files, we tend to use `"` instead of `'`

stephanecopin replied:
@fortmarek ah sorry about that, I have rubocop installed globally which do these changes automatically. I'll revert them. As an improvement, it may be good to such rules for Tuist so the style is kept consistency across all Ruby files?

fortmarek commented on projects/tuist/fixtures/app_with_test_plan/All.xctestplan:
Is this something we expect developers to be able to generate themselves? Does this PR provide value to tuist users until we deliver on generating the test plan functionality (https://github.com/tuist/tuist/discussions/5214)?

stephanecopin replied:
@fortmarek these will be generated by the user via Xcode. The PR allows to use test plans when using `tuist test` (which is not possible currently), so this is the value the PR provides, which is independent from being able to create test plans using Tuist.

fortmarek reviewed (commented):
Thanks a lot for this PR 👏 It's a great contribution

stephanecopin commented on Sources/TuistGraph/Models/TestPlan.swift:
```diff
@@ -1,12 +1,26 @@
 
-public struct TestPlan: Equatable, Codable {
+public struct TestPlan: Hashable, Codable {
+    public struct TestTarget: Hashable, Codable {
```
I'm fine with using it, reversing `isEnabled` with `!isSkipped`. I'll do the change

stephanecopin commented:
Hey @pepicrft @fortmarek @danyf90, I've finally got around to updating the PR, and should have addressed all of your comments. Please let me know if I missed anything!There was a bunch of additions, especially related to unit tests 😅
